bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/CaddxCommunicator.java

```diff
@@ -0,0 +1,521 @@
+                        logger.trace("CaddxCommunicator.run() Message expected. Message received");
+
+                        // Message expected. Message received.
+                        int receivedMessageType = incomingMessage.getMessageType();
+                        boolean isMessageExpected = IntStream.of(expectedMessageNumbers)
+                                .anyMatch(x -> x == receivedMessageType);
+
+                        if (!isMessageExpected) {
+                            logger.trace("Non expected message received exp:{}, recv: {}", expectedMessageNumbers,
+                                    receivedMessageType);
+
+                            // Non expected reply received
+                            if (outgoingMessage != null) {
+                                transmitFirst(outgoingMessage); // put message in queue again
+                                skipTransmit = true; // Skip the transmit on the next cycle to receive the panel message
+                            }
+                        }
+                    }
+                }
+
+                // Inform the listeners
+                if (incomingMessage != null && incomingMessage.isChecksumCorrect()) {
+                    for (CaddxPanelListener listener : listenerQueue) {
+                        listener.caddxMessage(this, incomingMessage);
+                    }
+                } else {
+                    logger.warn(
+                            "CaddxCommunicator.run() Received packet checksum does not match. in: {} {}, calc {} {}",
+                            incomingMessage.getChecksum1In(), incomingMessage.getChecksum2In(),
+                            incomingMessage.getChecksum1Calc(), incomingMessage.getChecksum2Calc());
+                }
+            }
+        } catch (IOException e) {
+            logger.debug("CaddxCommunicator.run() IOException. Stopping sender thread. {}", getSerialPortName());
+            Thread.currentThread().interrupt();
+        }
+
+        logger.warn("CaddxCommunicator.run() Sender thread stopped. {}", getSerialPortName());
+    }
+
+    /**
+     * Event handler to receive the data from the serial port
+     *
+     * @param SerialPortEvent serialPortEvent The event that occurred on the serial port
+     */
+    @Override
+    public void serialEvent(@Nullable SerialPortEvent serialPortEvent) {
+        if (serialPortEvent == null) {
+            return;
+        }
+
+        if (serialPortEvent.getEventType() == SerialPortEvent.DATA_AVAILABLE) {
+            logger.trace("Data receiving from the serial port");
+            if (protocol == CaddxProtocol.Binary) {
+                receiveInBinaryProtocol(serialPortEvent);
+            } else {
+                receiveInAsciiProtocol(serialPortEvent);
+            }
+        }
+    }
+
+    private int readByte(InputStream stream) {
+        try {
+            return stream.read();
+        } catch (IOException e) {
+            return -1;
```

Comment:
Instead of this method returning -1 to represent a terminating condition, use an exception instead since then they can be logged an handled appropriately by the callers of this method. Change this method to throw IOException and also throw a EOFException when read returns a -1. This will allow you to tell the difference between and end-of-file and a different error while at the same time handle them both as terminating conditions.

Reply:
Done